app/channel.go

```diff
@@ -2378,17 +2378,100 @@ func (a *App) MarkChannelAsUnreadFromPost(postID string, userID string) (*model.
 		return nil, err
 	}
 
+	threadId := post.RootId
+	if post.RootId == "" {
+		threadId = post.Id
+	}
+
 	unreadMentions, unreadMentionsRoot, err := a.countMentionsFromPost(user, post)
 	if err != nil {
 		return nil, err
 	}
 
-	if *a.Config().ServiceSettings.ThreadAutoFollow {
-		threadId := post.RootId
+	if !collapsedThreadsSupported {
+		// if root post,
+		// In CRT Supported Client: badge on channel only sums mentions in root posts including and below the post that was marked.
+		// In CRT Unsupported Client: badge on channel sums mentions in all posts (root & replies) including and below the post that was marked unread.
 		if post.RootId == "" {
-			threadId = post.Id
+			channelUnread, aErr := a.Srv().Store.Channel().UpdateLastViewedAtPost(post, userID, unreadMentions, unreadMentionsRoot, false, true)
+			if aErr != nil {
+				return channelUnread, model.NewAppError("MarkChannelAsUnreadFromPost", "app.channel.update_last_viewed_at_post.app_error", nil, aErr.Error(), http.StatusInternalServerError)
+			}
+
+			message := model.NewWebSocketEvent(model.WEBSOCKET_EVENT_POST_UNREAD, channelUnread.TeamId, channelUnread.ChannelId, channelUnread.UserId, nil)
```

Comment:
You know the "way we do things in server" much better than me, but would it make sense to extract the construction/sending of WS messages into a separate method make sense? These functions you're expanding are already getting very, very long.